### ISSUE TITLE
[HttpFoundation] Exception because HttpCache tries to access the container

### ISSUE THREAD

**Issue by aschempp** [symfony/symfony#40618]
**Symfony version(s) affected**: 5.2.6

**Description**
The changes in https://github.com/symfony/symfony/pull/40497 can cause an exception if the container is not yet bootet. As far as I understand the HttpCache concept, the kernel SHOULD not be bootet for performance reasons if the request is in the cache. Therefore, the changes in #40497 must be wrong.

**How to reproduce**

Here's the exception trace we're getting in a [Contao 4.11](https://contao.org) setup. We're also using the [FOSHttpCacheBundle](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle) and a custom entry point setup (not based on Symfony Flex)
 - [index.php](https://github.com/contao/contao/blob/4.11/manager-bundle/src/Resources/skeleton/web/index.php)
 - [ContaoCache](https://github.com/contao/contao/blob/18d4fa1de36eebe35b7572f3a5dffcd7f6b4090f/manager-bundle/src/HttpKernel/ContaoCache.php)
 - [EventDispatchingHttpCache](https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/SymfonyCache/EventDispatchingHttpCache.php)

```
PHP Fatal error:  Uncaught LogicException: Cannot retrieve the container from a non-booted kernel. in .../vendor/symfony/http-kernel/Kernel.php:307
Stack trace:
#0 .../vendor/symfony/framework-bundle/HttpCache/HttpCache.php(68): Symfony\Component\HttpKernel\Kernel->getContainer()
#1 .../vendor/friendsofsymfony/http-cache/src/SymfonyCache/EventDispatchingHttpCache.php(103): Symfony\Bundle\FrameworkBundle\HttpCache\HttpCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#2 .../web/index.php(31): Contao\ManagerBundle\HttpKernel\ContaoCache->handle(Object(Symfony\Component\HttpFoundation\Request))
#3 {main}
  thrown in .../vendor/symfony/http-kernel/Kernel.php on line 307
```

**Possible Solution**

As the HttpCache is imho not allowed to access container parameters, the only solution I can think of is keep behaviour in Symfony 4.4 and set `Request::enableHttpMethodParameterOverride();` in the entry point.

/cc @xabbuh

**Comment by nicolas-grekas** [comment]
Can you please confirm that #40619 fixes the issue ?

**Comment by aschempp** [comment]
Hmm, I fail to understand how you can configure the HttpCache in the container, if the instance is supposed to run _outside_ the kernel & container at all. Maybe @Toflar knows some details about that?

Not accessing `$this->kernel->getContainer()...` at runtime certainly fixes our issue though.

**Comment by Toflar** [comment]
I don't see how `HttpCache` can be configured via container either, that makes no sense to me. The whole point of it is to ensure Symfony doesn't need to be booted. But maybe there have been changes to that concept lately.

**Comment by nicolas-grekas** [comment]
You might have missed https://github.com/symfony/symfony/pull/37351
Fact is, both modes are supported since 5.2.

**Comment by Toflar** [comment]
Ah, indeed. Thanks for the hint :)

**Comment by aschempp** [comment]
Thanks for the PR link! Since the kernel seems now always bootet, shouldn't the `enableHttpMethodParameterOverride` always be called by the kernel instead? Looks like it might be called multiple times now if the request is not in cache?

**Comment by aschempp** [comment]
I mean, since the kernel is always bootet, https://github.com/symfony/symfony/blob/5.x/src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php#L95-L97 should always be called and the additional call shouldn't be necessary?

**Comment by nicolas-grekas** [comment]
#40497 fixes #40452, which explains why this is needed. TLDR, HttpCache is run before bundles are booted.

**Comment by aschempp** [comment]
Thanks for the quick fix @nicolas-grekas !
